### DISCUSSION TITLE
SIGSEGV with basic C++ program

### DISCUSSION THREAD

**Discussion by karimudev** [libvips/libvips#2357] [Q&A]
Hello,

I am trying to use libvips for a C++ image processing program, but for some reason I cannot get it to work even with a very basic program. The simple program that follows gives me a Segmentation Fault.

```c++
#include <vips/vips8>

using namespace vips;

int main(int argc, char **argv)
{
    if (VIPS_INIT(argv[0]))
        vips_error_exit(NULL);

    VImage in = VImage::new_from_file("Image.jpg");

    vips_shutdown();

    return 0;
}
```

The segmentation fault disappears if I remove the ```VImage``` assignment, or if I remove the ```vips_shutdown()``` call. I'm having the same issue with the [example C++ program](https://libvips.github.io/libvips/API/8.11/cpp/). GDB gives me the following backtrace for the Segfault.

```gdb
#0  0x00007ffff79f39b5 in g_mutex_lock () at /usr/lib/libglib-2.0.so.0
#1  0x00007ffff7c6812a in vips.argument_table_lookup () at /usr/lib/libvips.so.42
#2  0x00007ffff7c683b5 in vips_argument_map () at /usr/lib/libvips.so.42
#3  0x00007ffff7c684eb in  () at /usr/lib/libvips.so.42
#4  0x00007ffff7aa6591 in g_object_unref () at /usr/lib/libgobject-2.0.so.0
#5  0x0000555555555443 in vips::VObject::~VObject() ()
#6  0x000055555555545e in vips::VImage::~VImage() ()
#7  0x0000555555555337 in main ()
```

The whole library doesn't seem to work correctly on my system. I've tried to write another simple program that reads through a PNG's pixels one by one (without any kind of transformation or writing, only reading), but the program was extremely slow and seemed not to free any memory before I Ctrl+C'd it.

I'm using Arch Linux, and everything seems up-to-date on my system. Any idea as to what I can do to investigate the issue?

**Comment by jcupitt** [comment]
Hi @C0RNPOP,

What's happening is that vips is closing down in `vips_shutdown()`, but the destructor for `in` is running after that. Since vips has gone, you get a segv.

This works:

```C++
/* compile with
 *      g++ -g -Wall hello.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips8>

using namespace vips;

int main(int argc, char **argv)
{
    if (VIPS_INIT(argv[0]))
        vips_error_exit(NULL);

    {
        VImage in = VImage::new_from_file(argv[1]);
    }

    vips_shutdown();

    return 0;
}
```

Now the destructor runs before `vips_shutdown()`. Alternatively, you can simply remove the `vips_shutdown()` since libvips will run it for you automatically using `atexit()`. Some platforms don't have working atexit support, but you'll be fine on arch.

**Reply by karimudev** [reply]
Thank you, this has solved the issue.

**Comment by jcupitt** [comment]
Re. performance problems, usually you don't loop over pixels yourself, that will indeed be very slow, you use the libvips operations. 

What do you need to do? Maybe open a fresh discussion with some sample code.

**Reply by karimudev** [reply]
I am trying to cut any input image into small chunks, and then do some calculations over those chunks (like getting the average, or something similar).
However, I haven't tinkered a lot with the library nor checked the documentation in details (the only programs I wrote were small programs to test if everything worked correctly), so I'll continue trying a few stuff here and there. I'll open a discussion if I still have issues after trying a few stuff!
